--- Python-2.6.5/setup.py.orig	2014-05-15 12:16:04.753026020 -0500
+++ Python-2.6.5/setup.py	2014-05-15 12:24:29.245067782 -0500
@@ -1804,7 +1804,8 @@ class PyBuildExt(build_ext):
                 if not line:
                     ffi_inc = None
                     break
-                if line.startswith('#define LIBFFI_H'):
+                if line.startswith('#define LIBFFI_H') or \
+                   line.startswith('#define ffi_wrapper_h'):
                     break
         ffi_lib = None
         if ffi_inc is not None:
